src/btree/bt_compact.c

```diff
@@ -179,13 +179,24 @@ __compact_page(WT_SESSION_IMPL *session, WT_REF *ref, bool *skipp)
      * into memory while we were moving it.
      */
     if (previous_state == WT_REF_DISK && __wt_ref_addr_copy(session, ref, &copy)) {
-        bm = S2BT(session)->bm;
-        addr_size = copy.size;
-        WT_ERR(bm->compact_page_rewrite(bm, session, copy.addr, &addr_size, skipp));
-        if (!*skipp) {
-            copy.size = (uint8_t)addr_size;
-            WT_ERR(__compact_page_replace_addr(session, ref, &copy));
-            WT_STAT_DATA_INCR(session, btree_compact_pages_rewritten);
+        if (!ovfl_rewrite) {
+            bm = S2BT(session)->bm;
+            addr_size = copy.size;
+            WT_ERR(bm->compact_page_rewrite(bm, session, copy.addr, &addr_size, skipp));
+            if (!*skipp) {
+                copy.size = (uint8_t)addr_size;
+                WT_ERR(__compact_page_replace_addr(session, ref, &copy));
+                WT_STAT_DATA_INCR(session, btree_compact_pages_rewritten);
+            }
+        } else {
```

Comment:
@keithbostic this is based on your suggestion to bring pages in memory. Do you think if there is any side effects for this logic?

Reply:
Yes, I think this is close to the approach I'd suggest -- however, rather than doing an unlock/lock of the `WT_REF`, why not read the page into memory before acquiring the `WT_REF` lock in `__compact_page()`? That will ensure that `previous_state != WT_REF_DISK`, and then we will do our work in `__compact_page_inmem()`.

We will have to remember to release the page we read in as well.